File: 2020/go/day02/day02.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"regexp"
	"strconv"
)

func charcount(char, pwd string) (count int) {
	for _, c := range pwd {
		if string(c) == char {
			count += 1
		}
	}
	return
}

func main() {
	file, err := os.Open("/Users/hhh/JungleGym/advent_of_code/2020/inputs/day02.in")
	if err != nil {
		panic(err)
	}
	defer file.Close()

	validPwd := regexp.MustCompile(`(\d+)-(\d+) (\w): (\w+)`)
	scanner := bufio.NewScanner(file)
	p1, p2 := 0, 0

	for scanner.Scan() {
		line := scanner.Text()
		res := validPwd.FindAllStringSubmatch(line, -1)[0]
		min, _ := strconv.Atoi(string(res[1]))
		max, _ := strconv.Atoi(string(res[2]))
		char := res[3]
		pwd := res[4]

		// Part 1
		count := charcount(char, pwd)

		if count >= min && count <= max {
			p1 += 1
		}

		// Part 2
		c1 := string(pwd[min-1])
		c2 := string(pwd[max-1])
		if (c1 == char) && (c2 != char) {
			p2 += 1
		}
		if (c1 != char) && (c2 == char) {
			p2 += 1
		}

	}
	fmt.Printf("Day 1: %d\n", p1)
	fmt.Printf("Day 2: %d", p2)
}

File: 2020/day02.rb
input = File.readlines("/Users/hhh/JungleGym/advent_of_code/2020/inputs/day02.in").map { |line| line.strip }
valid_count = 0
p2_count = 0

input.each do |line|
  /(?<min>\d+)-(?<max>\d+) (?<char>\w): (?<pwd>\w+)/ =~ line
  if pwd.count(char).between? min.to_i, max.to_i
    valid_count += 1
  end

  i = min.to_i - 1
  j = max.to_i - 1

  if (pwd[i] == char) ^ (pwd[j] == char)
    p2_count += 1
  end

end

p "Part 1: #{valid_count}"
p "Part 2: #{p2_count}"
